rules/group.go

```diff
@@ -660,19 +662,32 @@ func (g *Group) RestoreForState(ts time.Time) {
 			continue
 		}
 
+		sset, err := alertRule.QueryforStateSeries(g.opts.Context, q)
+		if err != nil {
+			level.Error(g.logger).Log(
+				"msg", "Failed to restore 'for' state",
+				labels.AlertName, alertRule.Name(),
+				"stage", "Select",
+				"err", err,
+			)
+			continue
+		}
+
+		// No results for this alert rule.
+		if sset == nil {
+			level.Debug(g.logger).Log("msg", "Failed to find a series to restore the 'for' state", labels.AlertName, alertRule.Name())
+			continue
+		}
+
 		alertRule.ForEachActiveAlert(func(a *Alert) {
 			var s storage.Series
 
-			s, err := alertRule.QueryforStateSeries(g.opts.Context, a, q)
-			if err != nil {
-				// Querier Warnings are ignored. We do not care unless we have an error.
-				level.Error(g.logger).Log(
-					"msg", "Failed to restore 'for' state",
-					labels.AlertName, alertRule.Name(),
-					"stage", "Select",
-					"err", err,
-				)
-				return
+			// Find the series for the given alert from the set.
+			for sset.Next() {
```

Comment:
I don't fully understand how this works. Aren't we looking matches, given `sset` is never reset to "beginning" each when we look for each alert? How can we assume the alerts are iterated in the same order of the series in the response?

Reply:
You're right; thanks for pointing it out my silly mistake. It didn't work correctly to begin with, as I realised while fixing the tests. 

> How can we assume the alerts are iterated in the same order of the series in the response?

I think I get where you're going with this (but please correct me if I'm wrong). You want to co-relate the queried series and the "active alerts" by position to avoid the nested loops, but I don't think we can. There's no guarantee that the first evaluation after a restart (which is what we use to populate "active alerts") and whatever we have stored in `ALERT_FOR_SERIES` will be the same results.

That being said - I'm pretty inexperience when it comes to querying in Prometheus so if you have a better approach to do this I'm more than happy to implement it.

Reply:
>You want to co-relate the queried series and the "active alerts" by position to avoid the nested loops

Sorry for the misunderstanding. I wasn't suggesting doing it. My open question was "I think you're assuming they're in the same order but how is it possible?" but saying "we" instead of "you". Anyway, I see you've already changed the code and I think the approach you took (using a map) makes sense.